### PR TITLE
[SYCLomatic]Enable the migration of memoryClockRate and memoryBusWidth in cudaDeviceProp

### PR DESCRIPTION
**Pull request by wenhuiNi**
Signed-off-by: Ni, Wenhui <wenhui.ni@intel.com>

### CODE REVIEW

**Comment by zjin-lcf** [comment]
Do you know if the the namespace will remove "intel" (e.g. ext::intel::mem_clock_rate) in the future ?

**Comment by wenhuiNi** [comment]
@zjin-lcf 

> I don't know whether this namespace will be removed in the future.

BTW,this features(mem_clock_rate and memory_bus_width) is newly supported in intel/llvm.

File: clang/runtime/dpct-rt/include/device.hpp.inc [hunk 1]
```diff
@@ -630,6 +668,24 @@ public:
     prop.set_global_mem_size(
         get_info<sycl::info::device::global_mem_size>());
     prop.set_local_mem_size(get_info<sycl::info::device::local_mem_size>());
+    if (this->has(sycl::aspect::ext_intel_memory_clock_rate))
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/device.hpp.inc]: I think you need format this change, the code style is not aligned to the existing code.
"{" should not in a new line.

File: clang/runtime/dpct-rt/include/device.hpp.inc [hunk 2]
```diff
@@ -415,6 +431,26 @@ public:
     }
   }
 // DPCT_LABEL_END
+// DPCT_LABEL_BEGIN|device_info_set_memory_clock_rate|dpct
+// DPCT_PARENT_FEATURE|device_info
+// DPCT_DEPENDENCY_BEGIN
+// Device|device_info
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+  void set_memory_clock_rate(int memory_clock_rate){
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/device.hpp.inc]:
```suggestion
  void set_memory_clock_rate(int memory_clock_rate) {
```

File: clang/runtime/dpct-rt/include/device.hpp.inc [hunk 3]
```diff
@@ -415,6 +431,26 @@ public:
...
+    _memory_clock_rate = memory_clock_rate;
+  }
+// DPCT_LABEL_END
+// DPCT_LABEL_BEGIN|device_info_set_memory_bus_width|dpct
+// DPCT_PARENT_FEATURE|device_info
+// DPCT_DEPENDENCY_BEGIN
+// Device|device_info
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+  void set_memory_bus_width(int memory_bus_width){
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/device.hpp.inc]:
```suggestion
  void set_memory_bus_width(int memory_bus_width) {
```

File: clang/runtime/dpct-rt/include/device.hpp.inc (hunk 1 above)
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/device.hpp.inc]:
```suggestion
    if (this->has(sycl::aspect::ext_intel_memory_clock_rate)) {
      prop.set_memory_clock_rate(
          this->get_info<sycl::ext::intel::info::device::memory_clock_rate>());
    } else {
      prop.set_memory_clock_rate(0);
    }
    if (this->has(sycl::aspect::ext_intel_memory_bus_width)) {
      prop.set_memory_bus_width(
          this->get_info<sycl::ext::intel::info::device::memory_bus_width>());
    } else {
      prop.set_memory_bus_width(0);
    }
```

**Review by zhiweij1 — APPROVED** [review]

File: clang/runtime/dpct-rt/include/device.hpp.inc [hunk 4]
```diff
@@ -269,6 +269,22 @@ public:
 // DPCT_DEPENDENCY_END
 // DPCT_CODE
   size_t get_local_mem_size() const { return _local_mem_size; }
+// DPCT_LABEL_END
+// DPCT_LABEL_BEGIN|device_info_get_memory_clock_rate|dpct
+// DPCT_PARENT_FEATURE|device_info
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/device.hpp.inc]: Zhiwei's email says that: The “DPCT_PARENT_FEATURE” annotation is not recommended to use now.

File: clang/runtime/dpct-rt/include/device.hpp.inc [hunk 5]
```diff
@@ -631,6 +669,24 @@ public:
         get_info<sycl::info::device::global_mem_size>());
     prop.set_local_mem_size(get_info<sycl::info::device::local_mem_size>());
 
+#if(defined(SYCL_EXT_INTEL_DEVICE_INFO) && SYCL_EXT_INTEL_DEVICE_INFO >= 6)
+    if (this->has(sycl::aspect::ext_intel_memory_clock_rate)) {
+      prop.set_memory_clock_rate(
+          this->get_info<sycl::ext::intel::info::device::memory_clock_rate>());
+    } else {
+      prop.set_memory_clock_rate(0);
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/device.hpp.inc]: need set 0 here again?

**Review by tangjj11 — APPROVED** [review]
LGTM

File: clang/lib/DPCT/CallExprRewriter.cpp [hunk 6]
```diff
@@ -10,6 +10,7 @@
 #include "CallExprRewriterCommon.h"
 #include <memory>
 
+
```
Comment (tangjj11) [inline clang/lib/DPCT/CallExprRewriter.cpp]: why add this line?

**Review by ziranzha — APPROVED** [review]
LGTM

File: clang/test/dpct/helper_files_ref/include/device.hpp [hunk 7]
```diff
@@ -171,6 +179,8 @@ class device_info {
   int _minor;
   int _integrated = 0;
   int _frequency;
+  int _memory_clock_rate = 0;
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/device.hpp]: is int enough?  and non-zeor value may be better?

Comment (wenhuiNi) [reply clang/test/dpct/helper_files_ref/include/device.hpp]: 1. use unsigned int. 
         memoryClockRate and memoryBusWidth are both int type in cuda. https://docs.nvidia.com/cuda/cuda-runtime-api/structcudaDeviceProp.html.
         get_info<ext::intel::info::device::memory_bus_width>() will return a value of type uint32_t.
2. Use the value obtained on the gen9 platform as the default
        Device: Intel(R) UHD Graphics 630 [0x3e92]
        Memory clock rate: 0
        Memory bus width: 64

File: clang/runtime/dpct-rt/include/device.hpp.inc [hunk 8]
```diff
@@ -662,11 +662,28 @@ public:
     if (this->has(sycl::aspect::ext_intel_memory_clock_rate)) {
       prop.set_memory_clock_rate(
           this->get_info<sycl::ext::intel::info::device::memory_clock_rate>());
+    } else {
+      std::cerr
+          << "Query ext_intel_device_info_memory_clock_rate not supported "
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/device.hpp.inc]:
```suggestion
          << "Querying ext_intel_device_info_memory_clock_rate is not supported "
```

File: clang/runtime/dpct-rt/include/device.hpp.inc [hunk 9]
```diff
@@ -662,11 +662,28 @@ public:
...
+             "by the device"
+          << std::endl;
+      std::cerr << "memory_clock_rate default value is 0." << std::endl;
     }
     if (this->has(sycl::aspect::ext_intel_memory_bus_width)) {
       prop.set_memory_bus_width(
           this->get_info<sycl::ext::intel::info::device::memory_bus_width>());
+    } else {
+      std::cerr << "Query ext_intel_device_info_memory_bus_width not supported "
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/device.hpp.inc]:
```suggestion
      std::cerr << "Querying ext_intel_device_info_memory_bus_width is not supported "
```

File: clang/runtime/dpct-rt/include/device.hpp.inc [hunk 10]
```diff
@@ -662,11 +662,28 @@ public:
...
+                   "by the device"
+                << std::endl;
+      std::cerr << "memory_bus_width default value is 64." << std::endl;
     }
+#else
+    std::cerr << "get_device_info: query memory_clock_rate and "
+                 "memory_bus_width  is not supported by the compiler."
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/device.hpp.inc]:
```suggestion
                 "memory_bus_width  are not supported by the the compiler you currently used."
```

**Review by zhimingwang36 — CHANGES_REQUESTED** [review]

**Review by zhimingwang36 — APPROVED** [review]
LGTM

File: clang/test/dpct/helper_files_ref/include/device.hpp [hunk 11]
```diff
@@ -171,6 +179,8 @@ class device_info {
   int _minor;
   int _integrated = 0;
   int _frequency;
+  unsigned int _memory_clock_rate = 0;
```
Comment (YuriPlyakhin) [inline clang/test/dpct/helper_files_ref/include/device.hpp:182]: Would be good to add info on where are those default values coming from.

File: clang/test/dpct/helper_files_ref/include/device.hpp [hunk 12]
```diff
@@ -171,6 +179,8 @@ class device_info {
...
+  unsigned int _memory_bus_width = 64;
```
Comment (YuriPlyakhin) [inline clang/test/dpct/helper_files_ref/include/device.hpp:183]: Would be good to add info on where are those default values coming from.

File: clang/test/dpct/helper_files_ref/include/device.hpp [hunk 13]
```diff
@@ -276,6 +286,36 @@ class device_ext : public sycl::device {
...
+          this->get_info<sycl::ext::intel::info::device::memory_bus_width>());
+    } else {
+      std::cerr
+          << "Querying ext_intel_device_info_memory_bus_width is not supported "
+             "by the device"
+          << std::endl;
+      std::cerr << "memory_bus_width default value is 64." << std::endl;
+    }
+#else
+    std::cerr
```
Comment (YuriPlyakhin) [inline clang/test/dpct/helper_files_ref/include/device.hpp:311]: Can this be a compile-time warning?

Comment (wenhuiNi) [reply clang/test/dpct/helper_files_ref/include/device.hpp:311]: no, the hardware information is obtained at runtime

**Pull request merged**